src/statemachine/statemachine.js

```diff
@@ -0,0 +1,330 @@
+pc.extend(pc, (function () {
+
+
+    /**
+     * @callback
+     * @name pc.StateMachine~enter
+     * @description Called when a state is entered
+     * @param {String} previousState the state that is being switched from
+     * @param {String} state the current state
+     */
+
+    /**
+     * @callback
+     * @name pc.StateMachine~exit
+     * @description Called when a state is exited
+     * @param {String} newState the state that is being switched to
+     * @param {String} currentState the state that is being switched from
+     */
+
+    //For each loop
+    function forEach(collection, fn) {
+        if (Array.isArray(collection)) {
```

Comment:
According to this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray>

Array.isArray is compatible with everything:

Desktop <> Mobile <>
Feature	Chrome	Firefox (Gecko)	Internet Explorer	Opera	Safari
Basic support	5	4.0 <https://developer.mozilla.org/en-US/Firefox/Releases/4> (2.0)	9	10.5	5


> On 19 Jun 2015, at 11:27, moka <notifications@github.com> wrote:
> 
> In src/statemachine/statemachine.js <https://github.com/playcanvas/engine/pull/326#discussion_r32817834>:
> 
> > +     * @description Called when a state is entered
> > +     * @param {String} previousState the state that is being switched from
> > +     * @param {String} state the current state
> > +     */
> > +
> > +    /**
> > +     * @callback
> > +     * @name pc.StateMachine~exit
> > +     * @description Called when a state is exited
> > +     * @param {String} newState the state that is being switched to
> > +     * @param {String} currentState the state that is being switched from
> > +     */
> > +
> > +    //For each loop
> > +    function forEach(collection, fn) {
> > +        if (Array.isArray(collection)) {
> Array.isArray is part of ECMA6, is it polyfilled?
> Based on ECMA6 Compatibility Table <https://kangax.github.io/compat-table/es6/#Array_is_subclassable_Array.isArray_support> it is almost not implemented by other browsers.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/playcanvas/engine/pull/326/files#r32817834>.
> 


-- 
See our 2 minute video on DBS bringing gamification to Singapore here: 
www.3radical.com/resources/videos/#DBSVideo


3radical Limited, 1 Friary, Temple Quay, Bristol BS1 6EA, England.  
T: +44 117 344 5008 ¦ F : +44 117 344 5188

3radical Pte Ltd, 3 Church Street, Level 8 Samsung Hub, Singapore 049483.  
T: +65 6408 3711 ¦ F : +65 6408 0101

www.3radical.com

3radical Limited is a limited company registered in England and Wales 
under number 7872556.  The registered address of 3radical Limited is 1 
Friary, Temple Quay, Bristol BS1 6EA, England.

3radical Pte Ltd is a registered in Singapore under number 201312239K.  The 
registered address of 3radical Pte Ltd is 51 Goldhill Plaza, #07-10/11, 
Singapore 308900.

The information in this e-mail and any attachments may be confidential.  If 
you are not the intended recipient, please notify the sender as soon as 
possible and delete any copies. Unauthorised use, dissemination, 
distribution, publication or copying of this communication is prohibited.

Attachments to this email have been checked with virus detection 
software before transmission, however you should carry out virus checks 
before opening any attachment. 3radical Limited and 3radical Pte Ltd 
accepts no liability for any loss or damage which may be caused by software 
viruses.